### ISSUE TITLE
Chat input: adding character directly after mention causes changing mention to a pubkey

### ISSUE THREAD

**Issue by micieslak** [status-im/status-desktop#7686]
# Bug Report

## Description

Additionally, in some cases HTML snippet is displayed as well.

[Kazam_screencast_00024.webm](https://user-images.githubusercontent.com/20650004/193228349-484af806-f1e0-4c77-b085-2e1594ecda38.webm)

## Steps to reproduce

Insert mention, position cursor directly after mention using arrows, add some text, send

#### Expected behavior

Message should be displayed in the same way as in the input field.

#### Actual behavior

Mention is expanded to pubkey, in some cases html snippet is added

### Additional Information

- Status desktop version:
- Operating System:

**Comment by alexjba** [comment]
Status chat input has several issues that needs to be fixed and considered for the final solution that will fix this bug:

#### Functional issue
1. Mentions must end with one of the chars: isSpace(c) || c == '.' || c == ',' || c == ':' || c == ';'. In the UI it is possible to send mentions that won’t end with these chars. 

    Ex: Add mention, hit backspace, type text.<br>Result: As seen in this bug description - the message is malformed.
    
    Root cause: Wrong key/mouse event processing in StatusChatInput.qml. It should not let the user invalidate the mention without knowing. (UI shows the mention as being ok until the message is sent)<br>

2. Insert mention, insert mention, delete space, hit left, hit enter -> message is malformed<br>Root cause: duplicated mention in mentions array (StatusChatInput.qml)

3. Add mention, add text, hit backspace, place cursor at the end of a mention -> Mention is deleted.<br> Root cause: StatusChatInput onCursorPositionChanged processes text based on last known key used. In this case backspace, even if the cursor was moved by mouse.

4. MentionSuggestionPopup doesn’t work well with key navigation when StatusChatInput is used in edit message mode. To reproduce this send a message in community, try to edit, add @ to trigger the suggestions and navigate using up/down <br>Root cause: Key.up/down is processed on out of focus items

5. Add mention, click in the middle of the mention (cursor placed in the middle. Should not be possible since you cannot place the cursor in the middle by keyboard navigation), add character, the character is added immediately after mention (although the mention needs to end with space), the mention+new char is selected, hit enter -> message is malformed. <br> Root cause: moving cursor with mouse is not processed in the same way as moving cursor with keyboard nav

6. Create chat -> No public key conversion in CreateChatView.qml; no mentions possible also. <br>
Root cause: missing implementation

7. EmojiSuggestionPopup doesn’t work well with key navigation when StatusChatInput is used in edit message mode <br>Root cause: Key.up/down is processed on out of focus items (same as mentions)

8. Edit message -> No custom emoji support (needs to be parsed before showing the message input). <br>
Root cause: missing implementation

#### Design issues:
1. StatusChatInput: <br>
    a. Emoji and mentions are tightly coupled with the TextArea<br> As a result the implementation became sticky. It has high regression risk on every change<br>
    b. Redundant and also unused code. I guess this is a result of a).<br>
    c. Polluted public API. StatusChatInput API is used in different ways to obtain the same result (ex: Edit message feature and chat input feature). As a result we have different behaviours/bugs.
    
2. Duplicated code for message parsing: <br>Chat Root store, message store use duplicated code to parse the message (interpretMessage)<br>
MessageView.qml editCompletedHandler() similar parsing logic as sendMessage - just that it works better with mentions

My suggestion is to refactor StatusChatInput, to decouple emoji and mentions features and use them as decorators on top of TextArea. After this we can clean StatusChatInput, remove redundancy, unused code and we can clean the public API.
After refactoring, fixing all issues or adding other custom features would become much easier.

The result would look something like this:
```
//StatusChatInput.qml

Rectangle {
   [...]
   MentionsFeature {
    id: mentionsFeature
    wrapee: textArea
   }
   
   EmojiFeature {
    id: emojiFeature
    wrapee: textArea
   }
   
   TextArea {
    id: textArea
    [...]
    Key.forwardTo: [mentionsFeature, emojiFeature]
   }
}

//MentionsFeature.qml
Item {
    property TextArea wrapee
    Connections {
        target: wrapee
        onTextChanged: ...
        onCursorPositionChanged: ...
        [...]
    }
}
```

@noeliaSD, @micieslak, @caybro, @alexandraB99, @iurimatias.
I would appreciate some thoughts on this before I invest more time on it. <br>I'm barely scratching the surface of the chat input and I think we need to invest some time in refactoring this component, but I don't know if we have the time to do it, if it's planned already or if I would ovelap with someone else's work.

**Comment by caybro** [comment]
cc @PascalPrecht too since he's working on adding support for pasting images there

**Comment by alexjba** [comment]
I'll just add here the issues I find along the way so that we can have an overview of what needs to be fixed

#### Functional issue
9. Type a text like this: Just give me the http :D -> :D is not converted to emoji
Root cause: TextArea onTextChanged looks for http at [cursorPosition -7, cursorPosition] . If it contains http the emoji conversion will not happen. The assumption is that the word before cursor is http:/ or https:/

**Comment by noeliaSD** [comment]
We've already discussed in some point the idea of refactoring this component since it is complex and we are having bugs with it, I'd say, bc of this complexity. So, for me, it is a good idea to go for the refactor and try to split it in basic an easier components. 

I appreciate your detailed analysis and idea. :)

So, if we have the time to do it, better now than in some months. For me it is a yes, but @iurimatias knows the priorities!

**Comment by PascalPrecht** [comment]
I'd also like to see a workforce tackling StatusChatInput properly once and for all

**Comment by micieslak** [comment]
Thanks @alexjba for that analysis. It's definitely helpful. But before doing refactor here imo we should think about changing the approach. Because consecutive refactors in current form proved to be ineffective bc of many regressions. Therefore I would suggest setting up qml unit tests for that (for the whole project in general, storybook config is a good starting point for that probably) and try to do more test-driven development here. Of course decomposing into smaller parts is needed as all, defining interfaces of those components etc. But there are so many corner cases that without decent covering with automatic tests we are ineffective with development and bugfixing of that part of our app. Also using storybook for development could be useful.
While having tests we could start fixing bug from adding test covering that bug. And later fix that bug without risk of breaking functionality which is already covered. I know, it's obvious but at this stage of the project we definitely should start testing stuff.

**Comment by alexjba** [comment]
Good point @micieslak. Having decent test coverage would help a lot with the refactoring. My proposal here is to break it into several steps:
1. Add tests for currently working functionality
2. Fix all the issues and add tests for each fix
3. Once we have a decent test coverage we can start refactoring

Step 1 and 2 are interchangeable or can even be developed in parallel. But since all the issues here don't look critical (imo) for dogfooding I guess we can start by adding tests for the working functionality to lower the regression risk on the next step.
